Types for or/not/and, `ValidatorFn`, relax types (PR #903 + mods)

subvertallchris opened the pull request:
**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

I do not expect this to be a breaking change for TypeScript users but I think that there is still room for improvement.

**The PR fulfills these requirements:**

- [ ] It's submitted to the `dev` branch for v1.x (or to a previous version branch), _not_ the `master` branch
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] All tests are passing: https://github.com/vuelidate/vuelidate/blob/master/.github/CONTRIBUTING.md#development-setup
- [ ] New/updated tests are included

If adding a **new feature**, the PR's description includes:

- [x] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**

This is a fork of #903 which appears to resolve #917 with the changes discussed in the original PR's comments. The goal of this modded version of the original PR is to unblock teams like mine that need typed `vm` and are unable to use explicitly typed `ValidatorFn`.

I still have not figured out a good dev setup for this library so I unfortunately have not tested it; however, I do not expect any of the changes to be problematic as all they're doing is adding some optional types and adding some generics.

The original PR by @victorgarciaesgi says:

```
Fellowing previous PR, I added more Typescript upgrades for validators.

and: support for type casting
or: support for type casting
not: support for type casting
ValidatorFn: added vm: Vue as second param to access current instance
```

In the comments, we agreed that `ValidatorFn`'s `vm` should be typed as the union between _some component-specific state_ and `Vue`, which is one of the changes made here. The other changes, swapping out `unknown` for `any`, works around an issue wherein the library has types that reference other types which default to `unknown`, effectively preventing users from declaring `ValidatorFn` as described in [this comment here](https://github.com/vuelidate/vuelidate/pull/903#discussion_r710268529). I think that `unknown` is probably the correct type for `ValidatorFn` but if and only if `useVuelidate` is able to extract appropriate types and feed them down the pipe internally. I'd be happy to try this if someone can help me get a good TypeScript dev flow working.

@victorgarciaesgi, I hope I did not overstep by jumping on this, thank you for your work. 🙏  @basti-n maybe you can eyeball this and weigh in?

dobromir-hristov commented:
@subvertallchris thank you for your contribution. 

As an idea on how to test this, just run `npm run build` and then `npm pack` inside the `packages/vuelidate` folder.

Then you should be able to just install that into your test project using `npm install ../path/to/tar`.

subvertallchris commented:
@dobromir-hristov This is very helpful, thank you. I was able to confirm that it resolved my immediate issues around `ValidatorFn`, the type of `vm` and the issue caused by the default `unknown` generic param that would get passed around after calling `useVuelidate`. I'll have to poke around our app to see if there are components I can refactor quickly with rules that use `and`, `or, or `not`.

I'd also like to get another TypeScripter's opinion on the move from `unknown` to `any` since it's usually a bad idea, maybe someone has a better suggestion for the short-term.

subvertallchris commented:
@dobromir-hristov Any thoughts on this? I know I said I'd like another TypeScripter's opinion but if nobody volunteers, IMO it's worth moving ahead. They'll certainly make themselves known if there's a problem. 😈

dobromir-hristov commented on packages/validators/index.d.ts:
```diff
@@ -32,23 +32,23 @@ export const minLength: (
-  withParams: <T = unknown>(params: object, validator: ValidationRule<T>) => ValidationRuleWithParams
-  withMessage: <T = unknown>(message: string | ((params: MessageProps) => string), validator: ValidationRule<T>) => ValidationRuleWithParams
+  withParams: (params: object, validator: ValidationRule) => ValidationRuleWithParams
+  withMessage: (message: string | Function, validator: ValidationRule) => ValidationRuleWithParams
```
Why are we removing the more detailed declaration for this? Is it better to just have a function?

subvertallchris replied:
This change was part of the original PR, I think it was done before you updated the types in the main branch. I'm not sure what the intention was. I can't think of a time that I've ever used the `Function` keyword like that and don't want to speculate so I'm just going to update these to match main.

dobromir-hristov reviewed (commented):
Overall I think its fine. If no one else has any complaints, I can just merge it.

subvertallchris commented:
Cool, thanks. Not sure about your one question, can’t remember if I did
that or if that was in the original. I can check it with more context and
let you know on Monday. Thanks for checking in with this!

On Sat, Sep 25, 2021 at 4:05 PM Dobromir Hristov ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> Overall I think its fine. If no one else has any complaints, I can just
> merge it.
> ------------------------------
>
> In packages/validators/index.d.ts
> <https://github.com/vuelidate/vuelidate/pull/941#discussion_r716091760>:
>
> >  export const sameAs: <E = unknown>(
>    equalTo: E,
>    otherName?: string
>  ) => ValidationRuleWithParams<{ equalTo: E, otherName: string }>;
>  export const url: ValidationRuleWithoutParams;
>  export const helpers: {
> -  withParams: <T = unknown>(params: object, validator: ValidationRule<T>) => ValidationRuleWithParams
> -  withMessage: <T = unknown>(message: string | ((params: MessageProps) => string), validator: ValidationRule<T>) => ValidationRuleWithParams
> +  withParams: (params: object, validator: ValidationRule) => ValidationRuleWithParams
> +  withMessage: (message: string | Function, validator: ValidationRule) => ValidationRuleWithParams
>
> Why are we removing the more detailed declaration for this? Is it better
> to just have a function?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/vuelidate/vuelidate/pull/941#pullrequestreview-763626693>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA7IJ56B6VG3PY7O77QE3PDUDYTPZANCNFSM5ESG5YJA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

subvertallchris commented:
@dobromir-hristov I just modified the PR to fix those two function signatures that you mentioned in a comment. I rebased and forced push to my branch so your comment was wiped out but you should have received a github notification before that happened.

subvertallchris commented:
Thank you! 🙌 

Would you mind releasing a new alpha build?

dobromir-hristov commented:
Will do tonight yeah
